packages/test/src/test-workflows.ts

```diff
@@ -1544,4 +1571,120 @@ test('continue-as-new-same-workflow', async (t) => {
   }
 });
 
+test('patched:not-replay', async (t) => {
+  const { logs, script } = t.context;
+  {
+    const req = await activate(t, makeStartWorkflow(script));
+    compareCompletion(
+      t,
+      req,
+      makeSuccess([
+        makeSetChangeMarker('my-change-id', false),
+        makeStartTimerCommand({ timerId: '0', startToFireTimeout: msToTs(100) }),
+      ])
+    );
+  }
+  {
+    const req = await activate(t, makeFireTimer('0'));
+    compareCompletion(
+      t,
+      req,
+      makeSuccess([makeSetChangeMarker('my-change-id', false), makeCompleteWorkflowExecution()])
+    );
+  }
+  t.deepEqual(logs, [['has change'], ['has change 2']]);
+});
+
+test('patched:replay-no-marker', async (t) => {
+  const { logs, script } = t.context;
+  {
+    const req = await activate(t, makeStartWorkflow(script, undefined, undefined, true));
+    compareCompletion(
+      t,
+      req,
+      makeSuccess([
+        makeSetChangeMarker('my-change-id', false),
+        makeStartTimerCommand({ timerId: '0', startToFireTimeout: msToTs(100) }),
+      ])
+    );
+  }
+  {
+    const act = makeFireTimer('0');
+    act.isReplaying = true;
+    const req = await activate(t, act);
+    compareCompletion(
+      t,
+      req,
+      makeSuccess([makeSetChangeMarker('my-change-id', false), makeCompleteWorkflowExecution()])
+    );
+  }
+  t.deepEqual(logs, [['no change'], ['no change 2']]);
+});
+
+test('patched:replay-no-marker-then-not-replay', async (t) => {
+  const { logs, script } = t.context;
+  {
+    const req = await activate(t, makeStartWorkflow(script, undefined, undefined, true));
+    compareCompletion(
+      t,
+      req,
+      makeSuccess([
+        makeSetChangeMarker('my-change-id', false),
+        makeStartTimerCommand({ timerId: '0', startToFireTimeout: msToTs(100) }),
+      ])
+    );
+  }
+  // For this second activation we are no longer replaying, so in principle we could say we now
+  // do have the marker, but that would violate the guarantee that patch calls always return the
+  // same value, so we should still not have the change
+  {
+    const req = await activate(t, makeFireTimer('0'));
+    compareCompletion(
+      t,
+      req,
+      makeSuccess([makeSetChangeMarker('my-change-id', false), makeCompleteWorkflowExecution()])
+    );
+  }
+  t.deepEqual(logs, [['no change'], ['no change 2']]);
+});
+
+test('patched:replay-with-marker', async (t) => {
+  const { logs, script } = t.context;
+  {
+    const act = makeStartWorkflow(script, undefined, undefined, true);
+    act.jobs?.push(makeNotifyHasChangeJob('my-change-id'));
```

Comment:
Probably better to use [makeActivation](https://github.com/temporalio/sdk-node/blob/13d648ef929fac3571b15fce77aae09742c9c750/packages/test/src/test-workflows.ts#L116) and add a `makeStartWorkflowJob` helper which `makeStartWorkflow` can use.
As mentioned in another comment, I'd rather have core enforce that the hasChange notification jobs come first, they have to, always, in every SDK.